modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -1436,9 +1456,31 @@ def isin_func(df, values):
     convert_dtypes = Map.register(pandas.DataFrame.convert_dtypes)
     invert = Map.register(pandas.DataFrame.__invert__, dtypes="copy")
     isna = Map.register(pandas.DataFrame.isna, dtypes=np.bool_)
+    # `dtypes` cannot be naively hinted for these functions when called by the numpy API:
+    # if `where` and `out` are specified, then some values in the resulting dataframe
+    # may retain their original values, which would not have the bool datatype
```

Comment:
We can still naively hint dtypes here - we don't pass `out` to the algebra layer, and handle that at the API Layer instead (at the `fix_dtypes_and_determine_return` function), so the output of this will always be all bools - we only do the masking at the API Layer, and there's where we don't know what the output dtype will be.